queries/sun_position.py

```diff
@@ -0,0 +1,623 @@
+                )
+                sun_pos[_SOLAR_POSITIONS.MIÐNÆTTI] = solar_midnight
+
+                # Add solar positions for city on a specific date
+                date = datetime.date.today().replace(
+                    month=cast(int, month), day=int(sun_re.group(2))
+                )
+                data[city][date] = sun_pos
+
+    return data
+
+
+@cached(TTLCache(maxsize=1, ttl=86400))
+def _get_almanak_hi_data() -> _SOLAR_DICT_TYPE:
+    r = requests.get(_ALMANAK_HI_URL)
+    # Use beautiful soup to extract text from HTML response
+    # and split on newlines
+    text: List[str] = (
+        BeautifulSoup(r.text, "html.parser")
+        .get_text()
+        .replace("\r\n", "\n")
+        .split("\n")
+    )
+    return _parse_almanak_hi_data(text)
+
+
+def _find_closest_city(data: _SOLAR_DICT_TYPE, loc: LatLonTuple) -> Optional[str]:
+    """Find city closest to loc in data."""
+    closest_city = None
+    closest_distance = None
+
+    for city_name, city_dict in data.items():
+        dist = distance(loc, cast(LatLonTuple, city_dict["pos"]))
+
+        if closest_distance is None or dist < closest_distance:
+            closest_distance = dist
+            closest_city = city_name
+
+    return closest_city
+
+
+_SOLAR_ENUM_TO_WORD: Dict[_SOLAR_POS_ENUM, str] = {
+    _SOLAR_POSITIONS.MIÐNÆTTI: "Miðnætti",
+    _SOLAR_POSITIONS.DÖGUN: "Dögun",
+    _SOLAR_POSITIONS.BIRTING: "Birting",
+    _SOLAR_POSITIONS.SÓLRIS: "Sólris",
+    _SOLAR_POSITIONS.HÁDEGI: "Hádegi",
+    _SOLAR_POSITIONS.SÓLARLAG: "Sólarlag",
+    _SOLAR_POSITIONS.MYRKUR: "Myrkur",
+    _SOLAR_POSITIONS.DAGSETUR: "Dagsetur",
+    _SOLAR_POSITIONS.SÓLARHÆÐ: "Sólarhæð",
+}
+
+
+def _answer_closest_solar_data(
+    data: _SOLAR_DICT_TYPE, sun_pos: _SOLAR_POS_ENUM, qdate: datetime.date, city: str
+) -> AnswerTuple:
+    """
+    Create answer for a sun position in city/place on date qdate.
+    """
+    # Get closest date to qdate in data
+    closest_date: datetime.date = sorted(
+        (k for k in data[city] if isinstance(k, datetime.date)),
+        key=lambda d: abs(d - qdate),
+    )[0]
+
+    voice: str = ""
+    answer: str = ""
+
+    when: str
+    in_past: Optional[bool] = None
+    if qdate == datetime.date.today():
+        when = "í dag"
+    elif qdate == datetime.date.today() + datetime.timedelta(days=1):
+        when = "á morgun"
+        in_past = False
+    elif qdate == datetime.date.today() - datetime.timedelta(days=1):
+        when = "í gær"
+        in_past = True
+    else:
+        with changedlocale(category="LC_TIME"):
+            when = qdate.strftime("%-d. %B")
+            in_past = qdate < datetime.date.today()
+
+    if sun_pos == _SOLAR_POSITIONS.SÓLARHÆÐ:
+        if in_past is None:
+            is_will_was = "er"
+        else:
+            is_will_was = "var"
+
+        degrees = str(data[city][closest_date][sun_pos]).replace(".", ",")
+        answer = f"Sólarhæð um hádegi {when} {is_will_was} um {degrees} gráður."
```

Comment:
We have some nice logic somewhere to tell you whether to say "gráður" or "gráða" - i.e. "1 gráða", "21 gráða", "11 gráður".

Reply:
Good catch, fixed!